### PR TITLE
[SPARK-35258][SHUFFLE][YARN] Add new metrics to ExternalShuffleService for better monitoring

### PR DESCRIPTION
**Pull request by xkrogen**
### What changes were proposed in this pull request?
This adds two new additional metrics to `ExternalBlockHandler`:
- `blockTransferRate` -- for indicating the rate of transferring blocks, vs. the data within them
- `blockTransferAvgSize_1min` -- a 1-minute trailing average of block sizes transferred by the ESS

Additionally, this enhances `YarnShuffleServiceMetrics` to expose the histogram/`Snapshot` information from `Timer` metrics within `ExternalBlockHandler`.

### Why are the changes needed?
Currently `ExternalBlockHandler` exposes some useful metrics, but is lacking around metrics for the rate of block transfers. We have `blockTransferRateBytes` to tell us the rate of _bytes_, but no metric to tell us the rate of _blocks_, which is especially relevant when running the ESS on HDDs that are sensitive to random reads. Many small block transfers can have a negative impact on performance, but won't show up as a spike in `blockTransferRateBytes` since the sizes are small. Thus the new metrics to show information around average block size and block transfer rate are very useful to monitor the health/performance of the ESS, especially when running on HDDs. 

For the `YarnShuffleServiceMetrics`, currently the three `Timer` metrics exposed by `ExternalBlockHandler` are being underutilized in a YARN-based environment -- they are basically treated as a `Meter`, only exposing rate-based information, when the metrics themselves are collected detailed histograms of timing information. We should expose this information for better observability.

### Does this PR introduce _any_ user-facing change?
Yes, there are two entirely new metrics for the ESS, as documented in `monitoring.md`. Additionally in a YARN environment, `Timer` metrics exposed by the ESS will include more rich timing information.

### How was this patch tested?
New unit tests are added to verify that new metrics are showing up as expected.

We have been running this patch internally for approx. 1 year and have found it to be useful for monitoring the health of ESS and diagnosing performance issues.

### CODE REVIEW

**Comment by xkrogen** [comment]
cc @tgravescs @mridulm @dongjoon-hyun @gatorsmile

**Comment by HyukjinKwon** [comment]
cc @Ngone51 too FYI

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java [hunk 1]
```diff
@@ -98,10 +99,10 @@ public ExternalBlockHandler(
       OneForOneStreamManager streamManager,
       ExternalShuffleBlockResolver blockManager,
       MergedShuffleFileManager mergeManager) {
-    this.metrics = new ShuffleMetrics();
     this.streamManager = streamManager;
     this.blockManager = blockManager;
     this.mergeManager = mergeManager;
+    this.metrics = new ShuffleMetrics();
```
Comment (jaceklaskowski) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:105]: Why has this been moved down? Curious.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java [hunk 2]
```diff
@@ -291,9 +294,18 @@ public ShuffleMetrics() {
...
+        @Override
+        protected Ratio getRatio() {
+          return Ratio.of(
+              blockTransferRateBytes.getOneMinuteRate(),
+              blockTransferRate.getOneMinuteRate());
+        }
+      });
       allMetrics.put("registeredExecutorsSize",
-                     (Gauge<Integer>) () -> blockManager.getRegisteredExecutorsSize());
+          (Gauge<Integer>) blockManager::getRegisteredExecutorsSize);
```
Comment (jaceklaskowski) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:308]: Been a while since I was with Java, and I'm curious whether the change is going to work with Java 8 (that I assume Spark 3.x keeps supporting)?

File: docs/monitoring.md [hunk 3]
```diff
@@ -1361,7 +1361,9 @@ Note: applies when running in Spark standalone as worker
 ### Component instance = shuffleService
 Note: applies to the shuffle service
 
+- blockTransferRate (meter)
 - blockTransferRateBytes (meter)
+- blockTransferAvgTime_1min (gauge - 1 minute moving average)
```
Comment (jaceklaskowski) [inline docs/monitoring.md:1366]: nit: 1-minute (with the dash)

File: resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceMetricsSuite.scala [hunk 4]
```diff
@@ -37,28 +38,57 @@ class YarnShuffleServiceMetricsSuite extends SparkFunSuite with Matchers {
   val metrics = new ExternalBlockHandler(streamManager, blockResolver).getAllMetrics
 
   test("metrics named as expected") {
-    val allMetrics = Set(
+    val allMetrics = Seq(
```
Comment (jaceklaskowski) [inline resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceMetricsSuite.scala:41]: Why `Seq` here?

File: resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceSuite.scala [hunk 5]
```diff
@@ -403,7 +403,9 @@ class YarnShuffleServiceSuite extends SparkFunSuite with Matchers with BeforeAnd
     val metrics = metricSetRef.get(metricsSource).asInstanceOf[MetricSet].getMetrics
 
     assert(metrics.keySet().asScala == Set(
+      "blockTransferRate",
```
Comment (jaceklaskowski) [inline resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceSuite.scala:406]: Why not to use the same trick "// Use sorted Seq instead of Set for easier comparison when there is a mismatch" here?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java (hunk 1 above)
Comment (xkrogen) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:105]: Great question! Changing from:
```
      allMetrics.put("registeredExecutorsSize",
                     (Gauge<Integer>) () -> blockManager.getRegisteredExecutorsSize());
```
to
```
      allMetrics.put("registeredExecutorsSize",
          (Gauge<Integer>) blockManager::getRegisteredExecutorsSize);
```
Means that the `blockManager` object is evaluated at the time of `new ShuffleMetrics()`, instead of later at the time the metrics are gathered. When the instantiation is the first line, `blockManager` is still null. If it's at the end, `blockManager` has been initialized by then.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java (hunk 2 above)
Comment (xkrogen) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:308]: Yes, this is a Java 8 "method reference". Read more about them [here](https://docs.oracle.com/javase/tutorial/java/javaOO/methodreferences.html) (note that at the top of the page it says it is written against JDK 8).

File: resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceMetricsSuite.scala (hunk 4 above)
Comment (xkrogen) [reply resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceMetricsSuite.scala:41]: I think the comment below explains. You can't sort a `Set` so if we use `Set` here we have to call `toSeq.sorted` -- we might as well just use `Seq` to begin with.

File: resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceSuite.scala (hunk 5 above)
Comment (xkrogen) [reply resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceSuite.scala:406]: Updated here and in `ExternalShuffleServiceMetricsSuite`

**Comment by xkrogen** [comment]
Thanks for looking @jaceklaskowski ! Just put up a new commit addressing your comments.

**Comment by xkrogen** [comment]
@jaceklaskowski do you have any more comments here? Thanks a lot for your help so far!

**Comment by mridulm** [comment]
Is this dependent on [Ye's PR](https://github.com/apache/spark/pull/32007) @otterc ?

**Comment by otterc** [comment]
> Is this dependent on [Ye's PR](https://github.com/apache/spark/pull/32007) @otterc ?

No, it seems completely independent to me.

**Comment by mridulm** [comment]
Sorry my comment was for #32468 .. misposted it @otterc :-)

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java [hunk 6]
```diff
@@ -128,6 +142,12 @@ private static ShuffleServiceMetricsInfo getShuffleServiceMetricsInfoForCounter(
     return new ShuffleServiceMetricsInfo(name, "Value of counter " + name);
   }
 
+  private static ShuffleServiceMetricsInfo getMetricsInfoGenericValue(
```
Comment (dongjoon-hyun) [inline common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java:145]: Although this is a private function, to be consistent with the existing methods, `getMetricsInfoGenericValue` -> `getShuffleServiceMetricsInfoForValue` might be better.

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java [hunk 7]
```diff
@@ -78,7 +82,17 @@ public static void collectMetric(
           new ShuffleServiceMetricsInfo(name + "_rate1", "1 minute rate of timer " + name),
           t.getOneMinuteRate())
         .addGauge(new ShuffleServiceMetricsInfo(name + "_rateMean", "Mean rate of timer " + name),
-          t.getMeanRate());
+          t.getMeanRate())
+        .addGauge(getMetricsInfoGenericValue(timingName, "max"), snapshot.getMax())
+        .addGauge(getMetricsInfoGenericValue(timingName, "min"), snapshot.getMin())
+        .addGauge(getMetricsInfoGenericValue(timingName, "mean"), snapshot.getMean())
+        .addGauge(getMetricsInfoGenericValue(timingName, "stdDev"), snapshot.getStdDev());
+      for (int percentile : new int[] { 1, 5, 25, 50, 75, 95, 99 }) {
```
Comment (dongjoon-hyun) [inline common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java:90]: Is there a reason of this selection? In general, we use `Snapshot.get98thPercentile` and `Snapshot.get999thPercentile` additionally too.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java (hunk 2 above)
Comment (dongjoon-hyun) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:308]: Is there a reason why we touch another `registeredExecutorsSize` metric at a PR claiming to add new different metrics? Apparently, this looks like orthogonal to me and the PR description doesn't mention this at all either. If you need this for a different reason, shall we make a separate PR?

File: resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceMetricsSuite.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceMetricsSuite.scala:41]: This contribution is also orthogonal to the PR's main contribution but this looks okay in this case since it's a trivial.

File: resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceSuite.scala [hunk 8]
```diff
@@ -413,7 +416,7 @@ class YarnShuffleServiceSuite extends SparkFunSuite with Matchers with BeforeAnd
       "finalizeShuffleMergeLatencyMillis",
       "shuffle-server.usedDirectMemory",
       "shuffle-server.usedHeapMemory"
-    ))
+    ).sorted)
```
Comment (dongjoon-hyun) [inline resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceSuite.scala:419]: Ditto: Although this is an orthogonal contribution, it will be okay.

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java (hunk 7 above)
Comment (xkrogen) [reply common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java:90]: I added 98th and 99.9th percentiles. Thanks for the input!

One thing that may seem unusual is the low percentiles (1, 5, 25), which have found useful internally for a getting a better sense of the full range of timing information. Let me know if you think they are too non-standard to include here.

**Comment by xkrogen** [comment]
Thanks for the comments @dongjoon-hyun ! I have pushed up new commits addressing them.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java (hunk 1 above)
Comment (xkrogen) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java]: This change was reverted in a subsequent commit. Resolving.

**Comment by xkrogen** [comment]
@dongjoon-hyun or @jaceklaskowski -- any more comments here or are we good to go? I have other PRs waiting that depend on this which I would like to get posted if possible.

**Comment by xkrogen** [comment]
Rebased and reverted some unrelated changes, thanks to @otterc for the heads up there!

**Comment by xkrogen** [comment]
@dongjoon-hyun @jaceklaskowski gentle ping
or @mridulm if you want to take a look

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java [hunk 9]
```diff
@@ -78,7 +82,32 @@ public static void collectMetric(
...
+        String percentileStr;
+        switch (percentileThousands) {
+          case 10:
+            percentileStr = "1stPercentile";
+            break;
+          case 999:
+            percentileStr = "99.9thPercentile";
+            break;
+          default:
+            percentileStr = String.format("%d%sPercentile", percentileThousands / 10, "th");
```
Comment (mridulm) [inline common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java:104]: nit: move 'th' into the format ?

**Review by mridulm — APPROVED** [review]

**Comment by mridulm** [comment]
+CC @dongjoon-hyun for review.

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java (hunk 9 above)
Comment (xkrogen) [reply common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java:104]: Good catch, this was leftover from previous logic. thanks!

**Comment by dongjoon-hyun** [comment]
Sorry for being late, @mridulm and @xkrogen .

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java [hunk 10]
```diff
@@ -78,7 +82,32 @@ public static void collectMetric(
...
+        .addGauge(
+          getShuffleServiceMetricsInfoForGenericValue(timingName, "stdDev"), snapshot.getStdDev());
+      for (int percentileThousands : new int[] { 10, 50, 250, 500, 750, 950, 980, 990, 999 }) {
+        String percentileStr;
+        switch (percentileThousands) {
+          case 10:
+            percentileStr = "1stPercentile";
+            break;
+          case 999:
+            percentileStr = "99.9thPercentile";
```
Comment (dongjoon-hyun) [inline common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java:101]: Hi, @xkrogen . Thank you for adding this and I saw your [comment](https://github.com/apache/spark/pull/32388#discussion_r629507805).

Technically, `999thPercentile` is consistent with the conventional in Apache Spark instead of `99.9thPercentitle`. If you look at the other existing metrics, it has the following.
Screen Shot 2021-05-30 at 6 52 14 PM


And, just a question. May I ask where you borrow this `99.9thPercentitle`?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java [hunk 11]
```diff
@@ -264,6 +265,8 @@ private void checkAuth(TransportClient client, String appId) {
     private final Timer registerExecutorRequestLatencyMillis = new Timer();
     // Time latency for processing finalize shuffle merge request latency in ms
     private final Timer finalizeShuffleMergeLatencyMillis = new Timer();
+    // Block transfer rate in blocks per second
```
Comment (dongjoon-hyun) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:268]: Is this valid when we do `getContinuousBlocksData`?

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @xkrogen . I left two comments.

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java (hunk 10 above)
Comment (xkrogen) [reply common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleServiceMetrics.java:101]: Thanks for sharing, I did not realize there was an existing convention here. I've updated it to be `999thPercentile`. I didn't borrow it from anywhere in particular, just how I would say it aloud: "ninety-nine point ninth percentile".

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java (hunk 11 above)
Comment (xkrogen) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:268]: This is a great point, thanks Dongjoon!

Currently a single batch fetch will be considered as one block fetch by this metric, regardless of how many blocks are fetched in the block. For our purposes, this is what we're interested in, since a big part of what we want to understand with this is the number of random reads we're submitting.

However I see value in breaking it out as the actual number of blocks as well. Perhaps we can have two metrics, `blockTransferRate` and `blockFetchRequestRate`. In the case of non-batch fetches they will be the same, and with batch fetches the `blockTransferRate` will be higher than the `blockFetchRequestRate`. WDYT?

Comment (xkrogen) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:268]: On second thought, I don't like `blockFetchRequestRate` since I think `FetchRequest` sounds like it would be at the level of a `OpenBlocks`/`FetchShuffleBlocks` message vs. the individual response level.

Maybe `blockTransferMessageRate`? Or `blockBufferTransferRate`?

Comment (xkrogen) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:268]: Pushed up a new commit with `blockTransferMessageRate` to demonstrate the idea, let me know what you think.

Comment (xkrogen) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:268]: @dongjoon-hyun how does the new code look to you?

Comment (xkrogen) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:268]: @mridulm I would also be curious about your thoughts on the new metrics.

**Comment by xkrogen** [comment]
Rebased on `master` to resolve some conflicts with other shuffle metrics related changes.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java (hunk 11 above)
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:309]: In spark read side, we treat a single read of `ShuffleBlockBatchId` as a single block read (from metric point of view).
A [recent discussion](https://github.com/apache/spark/pull/32140#discussion_r652733447) of this in push based shuffle here for context (see `ShuffleBlockBatchId` in `ShuffleBlockFetcherIterator`).

Given that, I am fine with treating a batch read as a single read - given it would be contiguous (effectively, similar to reading a 'large' block).

Thoughts @dongjoon-hyun  ?

Comment (xkrogen) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:309]: Thanks for that context @mridulm! To clarify, are you suggesting we ditch the new proposal and only maintain the old version, which would treat a batch fetch as a single block transfer?

From the comments on #32140, it seems like @otterc was indicating that there's a future plan to add metrics on the push-based shuffle side that also break out the merged chunks in terms of number of underlying blocks. Should we do the same here (which is basically my new proposal)?

Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:309]: That would be a new metric, if introduced, in read side.
Currently, what is tracked treats it as a single request - that request could be a block fetch or a single fetch for a large block or merged block read.

Comment (xkrogen) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:309]: Okay. I am pretty neutral here, happy to go either way. @dongjoon-hyun or @otterc , any thoughts?

Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:309]: I don't have anything much to add here . Would be okay either ways.

Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:309]: Would definitely want to hear @dongjoon-hyun's thoughts here, given he has context !
Also, +CC @Ngone51 who is helping review the client side for push based shuffle and has looked at similar codepaths recently.

Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:309]: I am assuming there are no further suggestions to resolve here ... we can do a follow up in case there is a change in behavior required for 3.2
+CC @dongjoon-hyun

**Comment by mridulm** [comment]
Merging to master

**Comment by mridulm** [comment]
Thanks for working on this @xkrogen !
Thanks for the reviews @dongjoon-hyun, @otterc, @jaceklaskowski

**Comment by xkrogen** [comment]
Thanks @mridulm ! And many thanks to all reviewers, the feedback was much appreciated!
